File: content/snippets/go/interview_lcci_16_25_lru_cache.go
package main

import (
	"container/list"
	"fmt"
)

type node struct {
	key int
	val int
}

type LRUCache struct {
	cap  int
	list *list.List
	pos  map[int]*list.Element
}

func Constructor(capacity int) LRUCache {
	return LRUCache{cap: capacity, list: list.New(), pos: map[int]*list.Element{}}
}

func (c *LRUCache) Get(key int) int {
	if e, ok := c.pos[key]; ok {
		c.list.MoveToFront(e)
		return e.Value.(node).val
	}
	return -1
}

func (c *LRUCache) Put(key int, value int) {
	if e, ok := c.pos[key]; ok {
		e.Value = node{key: key, val: value}
		c.list.MoveToFront(e)
		return
	}
	if c.list.Len() == c.cap {
		tail := c.list.Back()
		delete(c.pos, tail.Value.(node).key)
		c.list.Remove(tail)
	}
	e := c.list.PushFront(node{key: key, val: value})
	c.pos[key] = e
}

func main() {
	cache := Constructor(2)
	cache.Put(1, 1)
	cache.Put(2, 2)
	fmt.Println(cache.Get(1))
	cache.Put(3, 3)
	fmt.Println(cache.Get(2))
	cache.Put(4, 4)
	fmt.Println(cache.Get(1))
	fmt.Println(cache.Get(3))
	fmt.Println(cache.Get(4))
}

File: content/snippets/cpp/interview_lcci_16_25_lru_cache.cpp
#include <iostream>
#include <list>
#include <unordered_map>
using namespace std;

class LRUCache {
public:
    explicit LRUCache(int capacity) : cap(capacity) {}

    int get(int key) {
        auto it = pos.find(key);
        if (it == pos.end()) return -1;
        items.splice(items.begin(), items, it->second);
        return it->second->second;
    }

    void put(int key, int value) {
        auto it = pos.find(key);
        if (it != pos.end()) {
            it->second->second = value;
            items.splice(items.begin(), items, it->second);
            return;
        }
        if (static_cast<int>(items.size()) == cap) {
            int old = items.back().first;
            items.pop_back();
            pos.erase(old);
        }
        items.push_front({key, value});
        pos[key] = items.begin();
    }

private:
    int cap;
    list<pair<int, int>> items;
    unordered_map<int, list<pair<int, int>>::iterator> pos;
};

int main() {
    LRUCache cache(2);
    cache.put(1, 1);
    cache.put(2, 2);
    cout << cache.get(1) << endl;
    cache.put(3, 3);
    cout << cache.get(2) << endl;
    cache.put(4, 4);
    cout << cache.get(1) << endl;
    cout << cache.get(3) << endl;
    cout << cache.get(4) << endl;
    return 0;
}
